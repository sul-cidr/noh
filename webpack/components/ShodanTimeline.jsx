/* eslint-disable array-callback-return */

import React, { Component } from "react";
import PropTypes from "prop-types";
import ShodanTimelineBlock from "./ShodanTimelineBlock";

class ShodanTimeline extends Component {
  constructor(props) {
    super(props);
    this.createSectionBlocks = this.createSectionBlocks.bind(this);
  }

  createSectionBlocks() {
    let position = 0;
    let currentWidth = 0;
    const sectionBlocks = [];
    this.props.sections.map(section => {
      // the key here has to change if name is not unique
      const newBlock = (
        <ShodanTimelineBlock
          key={section.sectionName.value}
          name={section.sectionName.value}
          url={section.sectionUrl}
          left={`${position}%`}
          maxIntensity={this.props.maxIntensity}
          intensity={section.intensity.number || "0"}
          // lack of start and end time data means duration isnt always being computed
          duration={section.endTime.value - section.startTime.value || 150}
          totalDuration={this.props.totalDuration}
        />
      );
      sectionBlocks.push(newBlock);
      // lack of start and end time data means duration isnt being computed
      const blockWidth =
        (section.endTime.value - section.startTime.value || 150) /
        this.props.totalDuration *
        100;
      currentWidth = blockWidth;
      position += currentWidth;
    });
    return sectionBlocks;
  }

  render() {
    const sectionBlocks = this.createSectionBlocks();
    return <div className="shodan-map"> {sectionBlocks}</div>;
  }
}

ShodanTimeline.propTypes = {
  sections: PropTypes.arrayOf(
    PropTypes.shape({
      sectionUrl: PropTypes.string,
      sectionName: PropTypes.shape({ value: PropTypes.string }),
      intensity: PropTypes.shape({ number: PropTypes.string }),
      startTime: PropTypes.shape({ value: PropTypes.number }),
      endTime: PropTypes.shape({ value: PropTypes.number })
    })
  ).isRequired,
  maxIntensity: PropTypes.number.isRequired,
  totalDuration: PropTypes.number.isRequired
};

export default ShodanTimeline;
