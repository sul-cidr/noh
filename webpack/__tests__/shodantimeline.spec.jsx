import React from "react";
import { shallow } from "enzyme";
import ShodanTimeline from "../components/ShodanTimeline";

describe("<ShodanTimeline>", () => {
  const sections = [
    {
      sectionName: { value: "Noriji" },
      sectionUrl: "/noriji",
      intensity: { number: "17" },
      startTime: { value: 0 },
      endTime: { value: 300 }
    },
    {
      sectionName: { value: "Maibataraki" },
      sectionUrl: "/maibataraki",
      intensity: { number: "12" },
      startTime: { value: 301 },
      endTime: { value: 1500 }
    }
  ];
  it("renders as expected", () => {
    const component = shallow(
      <ShodanTimeline
        sections={sections}
        maxIntensity={21}
        totalDuration={1800}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
