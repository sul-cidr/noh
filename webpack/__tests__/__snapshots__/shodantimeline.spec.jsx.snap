// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShodanTimeline> renders as expected 1`] = `
<div
  className="shodan-map"
>
   
  <ShodanTimelineBlock
    duration={300}
    intensity="17"
    key="Noriji"
    left="0%"
    maxIntensity={21}
    name="Noriji"
    totalDuration={1800}
    url="/noriji"
  />
  <ShodanTimelineBlock
    duration={1199}
    intensity="12"
    key="Maibataraki"
    left="16.666666666666664%"
    maxIntensity={21}
    name="Maibataraki"
    totalDuration={1800}
    url="/maibataraki"
  />
</div>
`;
